### PR TITLE
Add JWT authentication

### PR DESCRIPTION
**Pull request by lookuptable**
**Release note**:

```
Support node agent authentication using Google ID token.
```

### CODE REVIEW

File: cmd/node_agent/na/gcp_test.go [hunk 1]
```diff
@@ -36,10 +36,11 @@ func (fetcher *mockTokenFetcher) FetchToken() (string, error) {
 
 func TestGetDialOptions(t *testing.T) {
 	gcp := gcpPlatformImpl{&mockTokenFetcher{}}
+	config := &Config{RootCACertFile: "testdata/cert-chain-good.pem"}
+	options, _ := gcp.GetDialOptions(config)
```
Comment (wattli) [inline cmd/node_agent/na/gcp_test.go]: also check err is nill?

File: pkg/server/grpc/authenticator.go [hunk 2]
```diff
@@ -15,23 +15,43 @@
...
 	"golang.org/x/net/context"
 	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
 	"google.golang.org/grpc/peer"
 
 	"istio.io/auth/pkg/pki"
 )
 
+const (
+	httpAuthHeader    = "authorization"
```
Comment (wattli) [inline pkg/server/grpc/authenticator.go]: sort

File: pkg/server/grpc/authenticator.go [hunk 3]
```diff
@@ -58,5 +78,68 @@ func (cca *clientCertAuthenticator) authenticate(ctx context.Context) *user {
...
+
+	authHeader, exists := md[httpAuthHeader]
+	if !exists {
+		glog.Warning("no HTTP authorization header exists")
+		return ""
+	}
+
+	for _, value := range authHeader {
+		if strings.HasPrefix(value, bearerTokenPrefix) {
+			return strings.TrimPrefix(value, bearerTokenPrefix)
```
Comment (wattli) [inline pkg/server/grpc/authenticator.go]: Is it possible to have two bearer tokens?

Comment (lookuptable) [reply pkg/server/grpc/authenticator.go]: I feel this is rarely used. I'm looking into RFCs to see whether it's even allowed.

Comment (wattli) [reply pkg/server/grpc/authenticator.go]: Please ping back this thread once you are done.

Comment (lookuptable) [reply pkg/server/grpc/authenticator.go]: Dig more about this - though I can't find any RFC that forbids multiple authorization headers, the current grpc-go doesn't support multiple. See here: https://github.com/grpc/grpc-go/blob/master/transport/http2_client.go#L346

File: cmd/node_agent/na/gcp.go [hunk 4]
```diff
@@ -15,20 +15,28 @@
...
+const (
+	bearerTokenScheme = "Bearer"
+	httpAuthHeader    = "authorization"
+)
+
 type jwtAccess struct {
 	token string
 }
 
 func (j *jwtAccess) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
```
Comment (myidpt) [inline cmd/node_agent/na/gcp.go]: Where is this used?

File: pkg/server/grpc/authenticator.go (hunk 3 above)
Comment (wattli) [reply pkg/server/grpc/authenticator.go]: sg

File: pkg/server/grpc/server.go [hunk 5]
```diff
@@ -171,6 +180,15 @@ func (s *Server) applyServerCertificate() (*tls.Certificate, error) {
 	return &cert, nil
 }
 
+func (s *Server) authenticate(ctx context.Context) *user {
+	for _, authn := range s.authenticators {
+		if u := authn.authenticate(ctx); u != nil {
```
Comment (myidpt) [inline pkg/server/grpc/server.go]: I think you may want to be explicit about the order here. E.g., use client cert authentication if both cert and jwt are presented.

File: cmd/node_agent/na/gcp.go (hunk 4 above)
Comment (lookuptable) [reply cmd/node_agent/na/gcp.go]: This is used by gRPC internally when making a connection.

File: pkg/server/grpc/server.go (hunk 5 above)
Comment (lookuptable) [reply pkg/server/grpc/server.go]: Where/how do you want me to make it explicit? Adding comments to where the authenticators get initialized?

File: cmd/node_agent/na/gcp.go (hunk 4 above)
Comment (myidpt) [reply cmd/node_agent/na/gcp.go]: I see. Thanks.

File: pkg/server/grpc/server.go (hunk 5 above)
Comment (myidpt) [reply pkg/server/grpc/server.go]: Maybe make authenticators a map[int]authenticator ? The int index can be the authenticator type enum. And here you can do:
if s.authenticators[CertAuthenticator] != nil {
  s.authenticators[CertAuthenticator].authenticate(ctx)
  ...
}
if s.authenticators[JwtAuthenticator] != nil {
  s.authenticators[JwtAuthenticator].authenticate(ctx)
  ...
}

Comment (lookuptable) [reply pkg/server/grpc/server.go]: hmm I feel this is a verbose way to implement the same logic. if you're worrying about confusing users, I believe clear documentation makes more sense than code change.

Comment (myidpt) [reply pkg/server/grpc/server.go]: Sure, doc/comment is definitely helpful. We can use the order of s.authenticators to specify the authenticator usage order. So now it is first cert then JWT, as can be seen from the New() func.

**Pull request merged**